contracts/contracts/GrantRegistry.sol

```diff
@@ -1,135 +1,124 @@
-    grantRegistryId = _grantRegistryId;
-    metaPtr = _metaPtr;
-  }
-
-  // State
-  enum State {
-    CONFIG,
-    ACTIVE,
-    CLOSED
-  }
-  //State public state;
-
-  // Grant Pointer Object
-  // Mapping from GrantID to set of Owner Address, Payout Address, Metadata Pointer, Replaces Grant (optional)
-  // TODO: Implement safemath for avoiding overflows with ints
-  uint256 public grantCount = 0;
-  mapping(bytes32 => Grant) public grants;
-  bytes32[] public grantIds;
+  // --- Data ---
+  /// @notice Number of grants stored in this registry
+  uint96 public grantCount;
 
+  /// @notice Grant object
   struct Grant {
-    bytes32 id;
-    address owner;
-    address payout;
-    string metaPtr;
-    State state;
-    bytes32 replaces;
+    uint96 id; // grant ID, as uint96 to pack into same slot as owner (this implies a max of 2^96-1 = 7.9e28 grants)
+    address owner; // grant owner (has permissions to modify grant information)
+    address payee; // address that receives funds donated to this grant
+    string metaPtr; // URL pointing to grant metadata (for off-chain use)
   }
 
-  function activate(bytes32 _id) external onlyOwner {
-    grants[_id].state = State.ACTIVE;
-  }
+  /// @notice Mapping from Grant ID to grant data
+  mapping(uint96 => Grant) public grants; // TODO use an array instead with ID to index it? Which is better? Will array copy full array to memory in `getAllGrants`?
 
-  function closed(bytes32 _id) public onlyOwner {
-    grants[_id].state = State.CLOSED;
-  }
+  // --- Events ---
+  /// @notice Emitted when a new grant is created
+  event GrantCreated(uint96 indexed id, address indexed owner, address indexed payee, string metaPtr);
+
+  /// @notice Emitted when a grant's owner is changed
+  event GrantUpdated(uint96 indexed id, address indexed owner, address indexed payee, string metaPtr);
 
-  function isConfig(bytes32 _id) public view returns (bool) {
-    return grants[_id].state == State.CONFIG;
+  // --- Core methods ---
+  /**
+   * @notice Create a new grant in the registry
+   * @param _owner Grant owner (has permissions to modify grant information)
+   * @param _payee Address that receives funds donated to this grant
+   * @param _metaPtr URL pointing to grant metadata (for off-chain use)
+   */
+  function createGrant(
+    address _owner,
+    address _payee,
+    string memory _metaPtr
+  ) external {
+    uint96 _id = grantCount;
+    grants[_id] = Grant(_id, _owner, _payee, _metaPtr);
+    emit GrantCreated(_id, _owner, _payee, _metaPtr);
+    grantCount += 1;
   }
 
-  function isActive(bytes32 _id) public view returns (bool) {
-    return grants[_id].state == State.ACTIVE;
+  /**
+   * @notice Update the owner of a grant
+   * @param _id ID of grant to update
+   * @param _owner New owner address
+   */
+  function updateGrantOwner(uint96 _id, address _owner) external {
+    Grant storage grant = grants[_id];
+    require(msg.sender == grant.owner, "Not authorized");
```

Comment:
What's the actual gas savings from not using a modifier? Does the optimizer not inline modifiers?? The repeated code makes me a little twitchy but I do think keeping gas down is worthwhile.

Reply:
Here's my current understanding (have not tested/confirmed this recently): 

When using modifiers, the function body and modifier code are inlined together, so, compared to using a function to replace the modifier, a modifier will have bigger bytecode but will be cheaper since there's no JUMPs between methods. Functions (that replace a modifier's job) will have smaller bytecode but be a little more expensive due to JUMPs.

But in this case, our modifier would need to read from storage, AND our function needs to read from the same storage slot. I don't think there's a way to have the modifier and function use the same SLOAD (maybe the optimizer would do this? Haven't tested it), so inlining it this way ensures we only use one SLOAD

SLOADs [cost](https://github.com/wolflo/evm-opcodes/blob/hardfork-berlin/gas.md#a6-sload) 2100 gas for the first read of a slot and 100 gas for subsequent reads

Reply:
Actually we'd only read from the same storage slot twice in the case of updating `owner`, since `payee` and `metaPtr` have their own storage slots. So you could argue for only using this pattern when updating `owner`.

I think this pattern should still be the cheapest by a few gas though, so I think we should leave it for now and open a separate issue to gas-golf the whole contract more generally

Reply:
Gotcha. I think in general we should avoid over-optimization initially, favoring more readable code, and look to do a gas-golfing pass before deployment. This feels like it's bordering on that "premature optimization" line, since modifiers are a widely used language feature that improve readability and reduce the likelihood of a dev error. Let's leave it as-is for now, but be careful about that over optimization line moving forward!